### PR TITLE
update the API in the minimizer framework

### PR DESCRIPTION
**Pull request by yorkerlin**

### CODE REVIEW

File: src/shogun/optimization/GradientDescendUpdater.h [hunk 1]
```diff
@@ -95,6 +95,11 @@ class GradientDescendUpdater: public DescendUpdaterWithCorrection
 	virtual void update_variable(SGVector<float64_t> variable_reference,
 		SGVector<float64_t> raw_negative_descend_direction);
 
+
+	virtual LearningRate* get_learning_rate()
+	{
+		return m_learning_rate;
```
Comment (iglesias) [inline src/shogun/optimization/GradientDescendUpdater.h]: How is the memory management of the LearningRate* taken care of?

File: src/shogun/optimization/InverseScalingLearningRate.h [hunk 2]
```diff
@@ -60,14 +60,15 @@ class InverseScalingLearningRate: public LearningRate
 	virtual ~InverseScalingLearningRate() {}
 
 	/** Get the learning rate for descent direction
-	 * Note that every time this method is called, the call counter will increase by 1
-	 * Note that the learning rate usually is positive
+	 * @param with_update do we update the mutable variables
```
Comment (iglesias) [inline src/shogun/optimization/InverseScalingLearningRate.h]: Missing question mark, for consistency with the comment below.

File: src/shogun/optimization/GradientDescendUpdater.h (hunk 1 above)
Comment (yorkerlin) [inline src/shogun/optimization/GradientDescendUpdater.h]: Once we use C++11 compiler, we can use `std::shared_ptr`.
Last time, the buildbot (travis) complained about the `auto` keyword.
So, for now I use raw pointer and users have to new/delete pointers in the `main`.

File: src/shogun/optimization/ConstLearningRate.h [hunk 3]
```diff
@@ -62,12 +62,13 @@ class ConstLearningRate: public LearningRate
 	}
 
 	/** Get the learning rate for descent direction
+	 * @param with_update do we update the mutable variables?
 	 *
 	 * Note that the learning rate usually is positive
```
Comment (lisitsyn) [inline src/shogun/optimization/ConstLearningRate.h]: I think ```@param``` should come after ```Note ...```

**Comment by yorkerlin** [comment]
This update move the `learning rate pointer` from `Updater` to `Minimizer`.
For smooth penalty, using learning rate in Updater is fine since `gradient wrt penalty` is smooth. (some updater has its own build-in adaptive  learning rate)

For L1 penalty, we need an explicitly learning rate for `gradient wrt penalty`. 
Therefore, the `learning rate pointer` is moved into Minimizer.

**Comment by iglesias** [comment]
iirc we were already using C++11. @lisitsyn, @vigsterkr, is that right?

**Comment by yorkerlin** [comment]
@iglesias 
Traivs still complains about the `auto` keyword.
https://travis-ci.org/shogun-toolbox/shogun/jobs/75011240
https://travis-ci.org/shogun-toolbox/shogun/jobs/75014041

ref:
https://github.com/shogun-toolbox/shogun/pull/2876

**Comment by yorkerlin** [comment]
@iglesias 
@lisitsyn 
I merge this first.
Once the buildbot (travis) is compatible with C++11, I will use the the smart pointer.
